core/src/main/java/com/graphhopper/routing/ch/SplitArray2D.java

```diff
@@ -0,0 +1,124 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.ch;
+
+import java.util.Arrays;
+
+/**
+ * This is a more memory-efficient version of two equal length `ArrayList<T>[]`s, i.e. this is a fixed size array where
+ * each element is a variable sized sub-array and each sub-array is divided into two parts.
+ * This is more memory efficient than two arrays of `ArrayList`s, because it saves the object-overhead of using an
+ * ArrayList object for each sub-array.
+ * <p>
+ * The elements in each sub-array are divided into two parts with ranges [0,mid[ and [mid,size[. We can add elements
+ * to either of the two parts.
+ */
+class SplitArray2D<T> {
```

Comment:
Any idea for a better name? I am trying to say this is an array of arrays (2D) where each of the inner arrays is split into two parts (the elements of the two parts are separated from each other). But maybe not so important as this is only a package private class used by CHPreparationGraph...